### PR TITLE
Vend generated files to plugins

### PR DESCRIPTION
**Pull request by neonichu**
The package model vended to plugins currently does not contain any information about generated files which makes it difficult to work with plugins which depend on each other. It isn't quite clear to me how to fully rectify this situation since we do not have a full view of all generated files until after we executed all the plugins which is of course too late to vend that information to them. Possibly the API should be split in two where plugins first tell us about which files are generated and then the build commands are produced in a separate step?

Either way, I think we can mitigate the situation somewhat by improving the situation for a use case that basically already works: if multiple plugins are applied to the same target, later plugins should be able to know which files are generated by earlier plugins. The order of execution is already defined by the declaration in the package manifest.

As we are executing plugins, we can gather the generated files and vend them to plugins via a new API in `SourceModuleTarget`. Currently, this list will only be available when looking at the `target` instance passed as parameter, not the one in the graph, this should possibly be fixed.

rdar://103808529

### CODE REVIEW

File: Tests/SPMBuildCoreTests/PluginInvocationTests.swift [hunk 1]
```diff
@@ -188,15 +188,16 @@ class PluginInvocationTests: XCTestCase {
 
         // Construct a canned input and run plugins using our MockPluginScriptRunner().
         let outputDir = AbsolutePath("/Foo/.build")
-        let builtToolsDir = AbsolutePath("/Foo/.build/debug")
+        let builtToolsDir = AbsolutePath("/path/to/build/debug")
```
Comment (neonichu) [inline Tests/SPMBuildCoreTests/PluginInvocationTests.swift:191]: This is a consequence of using `mockBuildParameters` now.

File: Sources/PackagePlugin/PluginContextDeserializer.swift [hunk 2]
```diff
@@ -53,8 +53,8 @@ internal struct PluginContextDeserializer {
     /// Returns the `Target` that corresponds to the given ID (a small integer),
     /// or throws an error if the ID is invalid. The target is deserialized on-
     /// demand if it hasn't already been deserialized.
-    mutating func target(for id: WireInput.Target.Id) throws -> Target {
-        if let target = targetsById[id] { return target }
+    mutating func target(for id: WireInput.Target.Id, pluginGeneratedSources: [URL] = [], pluginGeneratedResources: [URL] = []) throws -> Target {
+        if let target = targetsById[id], target.sourceModule?.pluginGeneratedSources.count == pluginGeneratedSources.count, target.sourceModule?.pluginGeneratedResources.count == pluginGeneratedResources.count { return target }
```
Comment (neonichu) [inline Sources/PackagePlugin/PluginContextDeserializer.swift:57]: This seems a bit ugly.

**Comment by neonichu** [comment]
Need to add a test, but looking for some feedback on the general PR first:
- are we fine with the shape of the new public API that is being added to `PackagePlugin`
- are we fine with the general approach and with the limitations (generated files are only visible to other plugins applied to the same target via the `target` argument, not via the graph)

**Comment by neonichu** [comment]
@swift-ci please test

**Comment by neonichu** [comment]
Just removed some leftover print statements - no need to re-run CI for now since it initially passed, so the implementation is fine and I anyway need to add a test before landing this.

File: Sources/PackagePlugin/PluginContextDeserializer.swift (hunk 2 above)
Comment (MaxDesiatov) [reply Sources/PackagePlugin/PluginContextDeserializer.swift:57]: logic LGTM, but I see a usual formatting nit 😅

**Review by MaxDesiatov — APPROVED** [review]
Overall makes sense, but this needs tests before merging

**Comment by neonichu** [comment]
@swift-ci please test

**Comment by neonichu** [comment]
@swift-ci please test windows

**Comment by neonichu** [comment]
Some corelibs-foundation shenanigans:

```
[4/7] Compiling MyExecutable tool.swift
/tmp/Miscellaneous_Plugins_DependentPlugins.6SPuRm/Miscellaneous_Plugins_DependentPlugins/Sources/MyExecutable/tool.swift:15:24: error: no exact matches in call to initializer 
        """#.write(to: URL(filePath: path), atomically: true, encoding: .utf8)
                       ^
Foundation.URL:8:12: note: incorrect labels for candidate (have: '(filePath:)', expected: '(fileURLWithPath:)')
    public init(fileURLWithPath path: String)
           ^
Foundation.URL:2:12: note: incorrect labels for candidate (have: '(filePath:)', expected: '(fileReferenceLiteralResourceName:)')
    public init(fileReferenceLiteralResourceName name: String)
           ^
error: fatalError
```

**Comment by neonichu** [comment]
@swift-ci please test

**Comment by neonichu** [comment]
@swift-ci please test windows

**Comment by neonichu** [comment]
@swift-ci please test windows

**Pull request merged**